File: MobileApp/my_social_app/lib/models/notification.dart
import 'package:flutter/material.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:my_social_app/custom_packages/media/models/multimedia.dart';
import 'package:my_social_app/state/notifications_state.dart/notification_state.dart';
part 'notification.g.dart';

@immutable
@JsonSerializable()
class Notification{
  final int id;
  final DateTime createdAt;
  final int ownerId;
  final bool isViewed;
  final int type;
  final int userId;
  final String userName;
  final Multimedia? image;
  
  final int? questionId;
  final String? questionContent;
  final Multimedia? questionMedia;
  
  final int? commentId;
  final String? commentContent;

  final int? solutionId;
  final String? solutionContent;
  final Multimedia? solutionMedia;

  final int? repliedId;
  final String? repliedContent;

  final int? solutionVoteType;

  const Notification({
    required this.id,
    required this.ownerId,
    required this.createdAt,
    required this.isViewed,
    required this.type,
    required this.userId,
    required this.userName,
    required this.image,

    required this.questionId,
    required this.questionContent,
    required this.questionMedia,

    required this.commentId,
    required this.commentContent,

    required this.solutionId,
    required this.solutionContent,
    required this.solutionMedia,

    required this.repliedId,
    required this.repliedContent,

    required this.solutionVoteType
  });

  factory Notification.fromJson(Map<String, dynamic> json) => _$NotificationFromJson(json);
  Map<String, dynamic> toJson() => _$NotificationToJson(this);

  NotificationState toNotificationState()
    => NotificationState(
        id: id,
        ownerId: ownerId,
        createdAt: createdAt,
        isViewed: isViewed,
        userName: userName,
        type: type,
        userId: userId,
        image: image,

        questionId: questionId,
        questionContent: questionContent,
        questionMedia: questionMedia,

        commentId: commentId,
        commentContent: commentContent,

        solutionId: solutionId,
        solutionContent: solutionContent,
        solutionMedia: solutionMedia,

        repliedId: repliedId,
        repliedContent: repliedContent,

        solutionVoteType: solutionVoteType
      );
}

File: src/MySocailApp.Domain/NotificationDomain/NotificationAggregate/Entities/Notification.cs
using MySocailApp.Core;
using MySocailApp.Domain.NotificationDomain.NotificationAggregate.DomainEvents;
using MySocailApp.Domain.NotificationDomain.NotificationAggregate.ValueObjects;

namespace MySocailApp.Domain.NotificationDomain.NotificationAggregate.Entities
{
    public class Notification : Entity, IAggregateRoot
    {
        public int OwnerId { get; private set; }
        public NotificationType Type { get; private set; }
        public bool IsViewed { get; private set; }
        public int UserId { get; private set; }
        public string UserName { get; private set; }
        public Multimedia? Image { get; private set; }

        public int? QuestionId { get; private set; }
        public string? QuestionContent { get; private set; }
        public Multimedia? QuestionMedia { get; private set; }

        public int? CommentId { get; private set; }
        public string? CommentContent { get; private set; }

        public int? SolutionId { get; private set; }
        public string? SolutionContent { get; private set; }
        public Multimedia? SolutionMedia { get; private set; }

        private Notification() { }
        
        private Notification(int ownerId, NotificationType type, int userId, string userName, Multimedia? image)
        {
            OwnerId = ownerId;
            Type = type;
            UserId = userId;
            UserName = userName;
            Image = image;
        }

        public static Notification QuestionLikedNotification(int ownerId, int userId, string userName, Multimedia? image, int questionId,string? questionContent, Multimedia? questionMedia)
        {
            var notification = new Notification(ownerId, NotificationType.QuestionLikedNotification, userId, userName, image)
            {
                QuestionId = questionId,
                QuestionContent = questionContent,
                QuestionMedia = questionMedia,
                CreatedAt = DateTime.UtcNow,
            };
            notification.AddDomainEvent(new NotificationCreatedDomainEvent(notification));
            return notification;
        }

        public static Notification QuestionCommentCreatedNotification(int ownerId, int userId, string userName, Multimedia? image, int questionId, string? questionContent, Multimedia? questionMedia, int commentId, string commentContent)
        {
            var notification = new Notification(ownerId, NotificationType.QuestionCommentCreatedNotification, userId, userName, image)
            {
                QuestionId = questionId,
                QuestionContent = questionContent,
                QuestionMedia = questionMedia,
                CommentId = commentId,
                CommentContent = commentContent,
                CreatedAt = DateTime.UtcNow,
            };
            notification.AddDomainEvent(new NotificationCreatedDomainEvent(notification));
            return notification;
        }

        public static Notification SolutionCreatedNotification(int ownerId, int userId, string userName, Multimedia? image, int solutionId, string? solutionContent, Multimedia? solutionMedia) 
        {
            var notification = new Notification(ownerId, NotificationType.SolutionCreatedNotification, userId, userName, image)
            {
                CreatedAt = DateTime.UtcNow,
                SolutionId = solutionId,
                SolutionContent = solutionContent,
                SolutionMedia = solutionMedia
            };
            notification.AddDomainEvent(new NotificationCreatedDomainEvent(notification));
            return notification;
        }

        public static Notification SolutionCommentCreatedNotification(int ownerId, int userId, string userName, Multimedia? image, int solutionId, string? solutionContent, Multimedia? solutionMedia, int? commentId, string? commentContent)
        {
            var notification = new Notification(ownerId, NotificationType.SolutionCommentCreatedNotification, userId, userName, image)
            {
                CreatedAt = DateTime.UtcNow,
                SolutionId = solutionId,
                SolutionContent = solutionContent,
                SolutionMedia = solutionMedia,
                CommentId = commentId,
                CommentContent = commentContent,
            };
            notification.AddDomainEvent(new NotificationCreatedDomainEvent(notification));
            return notification;
        }

        public void MarkAsViewed() => IsViewed = true;

        public void UpdateUserName(string userName, DateTime updatedAt)
        {
            if (UpdatedAt >= updatedAt) return;

            UserName = userName;
            UpdatedAt = updatedAt;
        }

        public void UpdateImage(Multimedia? image,DateTime updatedAt)
        {
            if (UpdatedAt >= updatedAt) return;

            Image = image;
            UpdatedAt = updatedAt;
        }
    }
}
